adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,960 @@
+    socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (socket < 0) {
+        os_printf("failed\n");
+        goto failed3;
+    }
+    os_printf("OK\n");
+
+    os_printf("bind socket ......");
+    memset(&sock_addr, 0, sizeof(sock_addr));
+    sock_addr.sin_family = AF_INET;
+    sock_addr.sin_addr.s_addr = 0;
+    sock_addr.sin_port = htons(OPENSSL_LOCAL_TCP_PORT);
+    ret = bind(socket, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret) {
+        os_printf("failed\n");
+        goto failed3;
+    }
+    os_printf("OK\n");
+
+    os_printf("socket connect to remote ......");
+    memset(&sock_addr, 0, sizeof(sock_addr));
+    sock_addr.sin_family = AF_INET;
+    sock_addr.sin_addr.s_addr = target_ip.addr;
+    sock_addr.sin_port = htons(tls_io_instance->port);
+
+    httpconn_set_non_block(socket);
+
+    ret = connect(socket, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret == -1) {
+        ret = httpconn_net_errno(socket);
+        if (ret != EINPROGRESS){
+            os_printf("failed %s\n", tls_io_instance->hostname);
+            goto failed3;
+        } else{
+            os_printf("correct\n");
+        }
+    }
+    os_printf("OK\n");
+
+    char recv_buf[128];
+    int recv_bytes = sizeof(recv_buf);
+    for(;;){
+        FD_ZERO(&readset);
+        FD_SET(socket, &readset);
+    
+        FD_ZERO(&writeset);
+        FD_SET(socket, &writeset);
+    
+        FD_ZERO(&errset);
+        FD_SET(socket, &errset);
+    
+        ret = lwip_select(socket + 1, &readset, &writeset, &errset, NULL);
+        if (ret > 0){
+            if (FD_ISSET(socket, &writeset)){
+              break;
+            }
+    
+            if (FD_ISSET(socket, &readset)){
+                memset(recv_buf, 0, recv_bytes);
+                break;
+            }
+        }else{
+            os_printf("lwip_select %d\n", ret);
+        }
+    }
+
+    int s = socket;
+    ctx = SSL_CTX_new(TLSv1_client_method());
+    if (!ctx) {
+        os_printf("failed\n");
+
+    }
+    os_printf("OK\n");
+    
+    os_printf("SSL set fragment\n");
+    ret = SSL_set_fragment(ctx, OPENSSL_FRAGMENT_SIZE);
+
+    os_printf("SSL new \n");
+    ssl = SSL_new(ctx);
+
+    os_printf("SSL set fd\n");
+    SSL_set_fd(ssl, s);
+
+    os_printf("SSL select\n");
+
+    while (-1 == SSL_connect(ssl))
+    {  
+        FD_ZERO(&readset);
+        FD_SET(s, &readset);
+        FD_ZERO(&writeset);
+        FD_SET(s, &writeset);
+        FD_ZERO(&errset);
+        FD_SET(s, &errset);
+
+        ret = lwip_select(s + 1, &readset, &writeset, &errset, NULL);
+    }
+
+    printf("SSL ok\n");
+
+    tls_io_instance->ssl = ssl;
+    tls_io_instance->ssl_context = ctx;
+
+    
+    failed3:
+        //close(socket);
+        return ;
+}
+
+static int send_handshake_bytes(TLS_IO_INSTANCE* tls_io_instance)
+{
+    system_print_meminfo();
+    LogError("free heap size %d", system_get_free_heap_size());
+    openssl_thread_LWIP_CONNECTION(tls_io_instance);
```

Comment:
For item 1,  `tlsio_openssl_open` calls `send_handshake_bytes` which calls `openssl_thread_LWIP_CONNECTION`,  so this function is actually indirectly called by `tlsio_openssl_open`. Is this what you mean?